src/irr/asset/CGLSLBRDFBuiltinIncludeLoader.h

```diff
@@ -426,6 +465,43 @@ float irr_glsl_ggx_burley_aniso(float anisotropy, float a2, float TdotH, float B
 	return w2*w2*atab * irr_glsl_RECIPROCAL_PI;
 }
 
+//accounting for visibility of microfacets so basically this one should always be used because converges faster and doesnt "lose samples"
+//https://schuttejoe.github.io/post/ggximportancesamplingpart2/
+//again: everything happens in reflection space, so _V is supposed to be view vector in reflection space (geom normal is positive Z axis)
+//https://hal.archives-ouvertes.fr/hal-01509746/document
+//Also: problem is our anisotropic ggx ndf (above) has extremely weird API (anisotropy and a2 instead of ax and ay) and so it's incosistent with sampling function
+//Note: this is kinda blurry and i dont really understand whats going on here
+BSDFSample irr_glsl_ggx_cos_gen_sample(in uvec2 _sample, in vec3 _V, in float _ax, in float _ay)
+{
+    vec2 u = vec2(_sample)/float(4294967295);
+
+    // stretch view vector so that we're sampling as if roughness=1.0
+    vec3 V = normalize(vec3(_ax*_V.x, _ay*_V.y, _V.z));
+
+    //build orthonormal basis (V,t1,t2)
+    vec3 t1 = (V.z<0.9999) ? normalize(cross(V, vec3(0.0,0.0,1.0))) : vec3(0.0,0.0,1.0);
+    vec3 t2 = cross(t1,V);
+
+    //sample point with polar coords (phi,r)
+    float a = 1.0/(1.0 + V.z);    
+    float r = sqrt(u.x);//no idea why sqrt is here
```

Comment:
probably because stratified sampling preserves area, and area is proportional to radius squared

so if you want to map a linear change in area to a change in radius you need to sqrt the area (first random number == sub-circle area)

Reply:
yea, because without sqrt samples much more dence near to centre of disk, same thing as uniform disk sampling described in pbrt (not the concentric one)